# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: tools/profiler/core/platform.cpp
# Commit: 73ad83f70029
# Full Hash: 73ad83f7002982e37a4ae585a405466a601625ca
# Author: Gerald Squelart <gsquelart@mozilla.com>
# Date: 2020-08-20 09:32:09
# Regressor Bug: 1657174
# File Overlap Count: 1
# Description:
#   Bug 1657174 - Added more (un)registration TLS assertions - r=mstange
#   
#   Assertions are also clarified with messages, to better distinguish the same tests in different locations.
#   
#   Assertions should now cover all cases:
# ==============================================================================

diff -r 3f866663b0aa -r 73ad83f70029 tools/profiler/core/platform.cpp
--- a/tools/profiler/core/platform.cpp	Wed Aug 19 22:59:22 2020 +0000
+++ b/tools/profiler/core/platform.cpp	Wed Aug 19 17:49:12 2020 +0000
@@ -3848,6 +3848,12 @@
     }
   }
 
+  MOZ_RELEASE_ASSERT(TLSRegisteredThread::RegisteredThread(aLock),
+                     "TLS should be set when registering thread");
+  MOZ_RELEASE_ASSERT(
+      registeredThread == TLSRegisteredThread::RegisteredThread(aLock),
+      "TLS should be set as expected when registering thread");
+
   ProfilingStack* profilingStack =
       &registeredThread->RacyRegisteredThread().ProfilingStack();
 
@@ -5081,6 +5087,12 @@
         "profiler_register_thread again",
         TextMarkerPayload(text, TimeStamp::NowUnfuzzed()), &lock);
 
+    MOZ_RELEASE_ASSERT(TLSRegisteredThread::RegisteredThread(lock),
+                       "TLS should be set when re-registering thread");
+    MOZ_RELEASE_ASSERT(
+        thread == TLSRegisteredThread::RegisteredThread(lock),
+        "TLS should be set as expected when re-registering thread");
+
     return &thread->RacyRegisteredThread().ProfilingStack();
   }
 
@@ -5103,11 +5115,14 @@
   // We don't call RegisteredThread::StopJSSampling() here; there's no point
   // doing that for a JS thread that is in the process of disappearing.
 
-  RegisteredThread* registeredThread = FindCurrentThreadRegisteredThread(lock);
-  if (registeredThread) {
-    MOZ_RELEASE_ASSERT(TLSRegisteredThread::RegisteredThread(lock));
-    MOZ_RELEASE_ASSERT(registeredThread ==
-                       TLSRegisteredThread::RegisteredThread(lock));
+  if (RegisteredThread* registeredThread =
+          FindCurrentThreadRegisteredThread(lock);
+      registeredThread) {
+    MOZ_RELEASE_ASSERT(TLSRegisteredThread::RegisteredThread(lock),
+                       "TLS should be set when un-registering thread");
+    MOZ_RELEASE_ASSERT(
+        registeredThread == TLSRegisteredThread::RegisteredThread(lock),
+        "TLS should be set as expected when un-registering thread");
     RefPtr<ThreadInfo> info = registeredThread->Info();
 
     DEBUG_LOG("profiler_unregister_thread: %s", info->Name());
@@ -5121,12 +5136,14 @@
     // thread is unregistering itself and won't need the ProfilingStack anymore.
     TLSRegisteredThread::ResetRegisteredThread(lock);
     TLSRegisteredThread::ResetAutoProfilerLabelProfilingStack(lock);
-    MOZ_RELEASE_ASSERT(!TLSRegisteredThread::RegisteredThread(lock));
 
     // Remove the thread from the list of registered threads. This deletes the
     // registeredThread object.
     CorePS::RemoveRegisteredThread(lock, registeredThread);
     MOZ_RELEASE_ASSERT(!FindCurrentThreadRegisteredThread(lock));
+    MOZ_RELEASE_ASSERT(
+        !TLSRegisteredThread::RegisteredThread(lock),
+        "TLS should have been reset after un-registering thread");
   } else {
     LOG("profiler_unregister_thread() - thread %d already unregistered",
         profiler_current_thread_id());
@@ -5150,7 +5167,9 @@
     //   (Whether or not it should, this does happen in practice.)
     //
     // Either way, TLSRegisteredThread should be empty.
-    MOZ_RELEASE_ASSERT(!TLSRegisteredThread::RegisteredThread(lock));
+    MOZ_RELEASE_ASSERT(
+        !TLSRegisteredThread::RegisteredThread(lock),
+        "TLS should have been reset when thread was previously un-registered");
   }
 }
 
